### PR TITLE
fix: SQL Injection via dot-notation sub-key name in `Increment` operation on PostgreSQL (GHSA-gqpp-xgvh-9h7h)

### PR DESCRIPTION
**Pull request by mtrezza**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).

## Issue

SQL Injection via dot-notation sub-key name in `Increment` operation on PostgreSQL ([GHSA-gqpp-xgvh-9h7h](https://github.com/parse-community/parse-server/security/advisories/GHSA-gqpp-xgvh-9h7h))

## Tasks

- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)
- [ ] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK 

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed a PostgreSQL issue where Increment operations with dot-notation sub-keys could be misinterpreted; field names with special characters are now handled safely to prevent unintended SQL behavior.

* **Tests**
  * Added a PostgreSQL-only test suite covering multiple dot-notation sub-key name edge cases (single/double-quote and valid JSONB forms) plus a normal nested Increment verification.

### CODE REVIEW

**Comment by parseplatformorg** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/acinader/pr-checks/37d0481b-a1a3-4670-8dbc-6bd235e5bf48?refs=3f0755f1386b3a7f16ab991e3e729ae4896f2621&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 1]
```diff
@@ -208,6 +208,8 @@ const handleDotFields = object => {
   return object;
 };
 
+const escapeSqlString = value => value.replace(/'/g, "''");
+
```
[the remark this answers is not in the document]
Comment (mtrezza) [reply src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:212]: This is a data integrity observation, not a security issue. Both interpolation sites for the sub-key sit inside single-quoted SQL string literals. In PostgreSQL, only `'` can terminate a single-quoted string, `"` and `\` have no special SQL meaning in that context, so they cannot enable SQL injection.

What `"` can do is produce malformed JSONB inside the `CONCAT(...)::jsonb` expression, which causes a database error. This is the same behavior you would get from inserting any structurally invalid data — an error, not an exploit.

We also verified that a `"` payload crafted to produce *valid* JSONB (e.g. `stats.x":0,"pg_sleep(3)`) still cannot execute arbitrary SQL, since the SQL string boundary is never breached. Tests for both scenarios are included in this PR.

Adding JSON escaping here would silently accept sub-keys containing `"` and `\` by transforming them, which changes the semantics of what the client sent. The current behavior (rejecting with an error) is arguably more correct for this edge case. Addressing JSON-level key escaping is a separate concern from this security fix and would be tracked separately if needed.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [9.6.0-alpha.5](https://github.com/parse-community/parse-server/releases/tag/9.6.0-alpha.5)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [9.6.0](https://github.com/parse-community/parse-server/releases/tag/9.6.0)
